collection.go

```diff
@@ -336,12 +404,85 @@ func (c *Collection) Query(ctx context.Context, queryText string, nResults int,
 		return nil, errors.New("queryText is empty")
 	}
 
-	queryVectors, err := c.embed(ctx, queryText)
+	queryVector, err := c.embed(ctx, queryText)
 	if err != nil {
 		return nil, fmt.Errorf("couldn't create embedding of query: %w", err)
 	}
 
-	return c.QueryEmbedding(ctx, queryVectors, nResults, where, whereDocument)
+	return c.QueryEmbedding(ctx, queryVector, nResults, where, whereDocument)
+}
+
+// Performs an exhaustive nearest neighbor search on the collection.
+//
+//   - options: The options for the query. See QueryOptions for more information.
+func (c *Collection) QueryWithOptions(ctx context.Context, options QueryOptions) ([]Result, error) {
+	if options.QueryText == "" && len(options.QueryEmbedding) == 0 {
+		return nil, errors.New("QueryText and QueryEmbedding options are empty")
+	}
+
+	var err error
+	queryVector := options.QueryEmbedding
+	if len(queryVector) == 0 {
+		queryVector, err = c.embed(ctx, options.QueryText)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't create embedding of query: %w", err)
+		}
+	}
+
+	negativeMode := options.NegativeMode
+	if negativeMode == "" {
+		negativeMode = NEGATIVE_MODE_SUBTRACT
```

Comment:
Your original PR implementation was with the subtract because it was simple to implement. Now that you've implemented all three (:muscle:!), isn't the filter one the best?

For the subtract my understanding is that it can lead to the following: You do a similarity search among product reviews, and when using "it's great" as query, "it's bad" as negative query, it can end up finding all "it's so-so" reviews. What I assume is more common is to use the filter to make sure no negative review is part of the results, in case they slipped through due to some additional words in the review that somehow make it similar to the positive one in the eyes of the embedding model.

Or is my understanding wrong?

An alternative could be to not define a default and always require the caller to pass an explicit value.

Reply:
I removed the default mode so you're force to pick one now. I agree that for most the filter mode would make most sense, I have put that on top of the constants.

I have also reworked the filtering code to return more results. Instead of filtering the N results you get back, it now filters at query time to still try to guarantee that you get N results (unless it filters more than that).